File: app/src/main/java/com/simon/harmonichackernews/ui/editor/ComposeEditorCoordinator.java
package com.simon.harmonichackernews.ui.editor;

import android.os.Bundle;
import android.text.TextUtils;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.simon.harmonichackernews.MainActivity;
import com.simon.harmonichackernews.R;
import com.simon.harmonichackernews.network.UserActions;
import com.simon.harmonichackernews.ui.common.CaptchaResultCallback;
import com.simon.harmonichackernews.utils.Utils;

import okhttp3.Response;

/** Owns submission side effects while the editor itself is a MainActivity Compose destination. */
public final class ComposeEditorCoordinator {
    private final MainActivity activity;
    private final Runnable onFinished;
    private final int id;
    private final int type;
    private final int titleMaxLength;
    private final String parentText;
    private final String postTitle;
    private final String user;
    private boolean submitting;
    private ComposeEditorController controller;

    public ComposeEditorCoordinator(
            @NonNull MainActivity activity,
            @NonNull Bundle arguments,
            @NonNull Runnable onFinished) {
        this.activity = activity;
        this.onFinished = onFinished;
        id = arguments.getInt(ComposeEditorContract.EXTRA_ID, -1);
        type = arguments.getInt(
                ComposeEditorContract.EXTRA_TYPE,
                ComposeEditorContract.TYPE_POST);
        parentText = arguments.getString(ComposeEditorContract.EXTRA_PARENT_TEXT);
        postTitle = arguments.getString(ComposeEditorContract.EXTRA_POST_TITLE);
        user = arguments.getString(ComposeEditorContract.EXTRA_USER);
        titleMaxLength = activity.getResources().getInteger(R.integer.title_max_length);
    }

    public int getType() {
        return type;
    }

    public int getTitleMaxLength() {
        return titleMaxLength;
    }

    @Nullable
    public String getParentText() {
        return parentText;
    }

    @Nullable
    public String getPostTitle() {
        return postTitle;
    }

    @Nullable
    public String getUser() {
        return user;
    }

    public void attachController(@NonNull ComposeEditorController controller) {
        this.controller = controller;
        controller.setSubmitting(submitting);
    }

    public void submit(@NonNull ComposeEditorSubmission submission) {
        submitValues(
                submission.getTitle(),
                submission.getUrl(),
                submission.getText(),
                submission.getComment());
    }

    private void submitValues(
            @NonNull String submittedTitle,
            @NonNull String submittedUrl,
            @NonNull String submittedText,
            @NonNull String submittedComment) {
        if (submitting) return;

        if (type == ComposeEditorContract.TYPE_POST) {
            if (TextUtils.isEmpty(submittedTitle)
                    || submittedTitle.length() > titleMaxLength
                    || (TextUtils.isEmpty(submittedUrl) && TextUtils.isEmpty(submittedText))) {
                return;
            }
        } else if (TextUtils.isEmpty(submittedComment)) {
            return;
        }

        if (!Utils.isNetworkAvailable(activity)) {
            showSubmissionFailure(
                    null,
                    null,
                    type == ComposeEditorContract.TYPE_POST ? null : submittedComment);
            return;
        }

        setSubmitting(true);
        if (type == ComposeEditorContract.TYPE_POST) {
            submitPost(submittedTitle, submittedText, submittedUrl);
        } else {
            submitComment(submittedComment);
        }
    }

    private void submitPost(@NonNull String title, @NonNull String text, @NonNull String url) {
        UserActions.submit(title, text, url, activity, new UserActions.ActionCallback() {
            @Override
            public void onSuccess(Response response) {
                Toast.makeText(
                        activity,
                        "Post submitted, it might take a minute to show up",
                        Toast.LENGTH_SHORT).show();
                onFinished.run();
            }

            @Override
            public void onFailure(String summary, String response) {
                setSubmitting(false);
                showSubmissionFailure(summary, response, null);
            }

            @Override
            public void onCaptchaRequired(UserActions.CaptchaChallenge challenge) {
                UserActions.ActionCallback callback = this;
                activity.showCaptchaDialog(
                        challenge,
                        new CaptchaResultCallback() {
                            @Override
                            public void onCaptchaResponse(
                                    UserActions.CaptchaChallenge captchaChallenge,
                                    String captchaResponse) {
                                if (captchaChallenge.isLoginChallenge()) {
                                    UserActions.submitAfterLoginCaptcha(
                                            title,
                                            text,
                                            url,
                                            activity,
                                            captchaChallenge,
                                            captchaResponse,
                                            callback);
                                } else {
                                    UserActions.continueCaptchaAction(
                                            activity,
                                            captchaChallenge,
                                            captchaResponse,
                                            callback);
                                }
                            }

                            @Override
                            public void onCaptchaCancelled() {
                                setSubmitting(false);
                                Toast.makeText(
                                        activity,
                                        "Post submission requires completing the HN captcha",
                                        Toast.LENGTH_SHORT).show();
                            }
                        });
            }
        });
    }

    private void submitComment(@NonNull String commentText) {
        UserActions.comment(String.valueOf(id), commentText, activity,
                new UserActions.ActionCallback() {
                    @Override
                    public void onSuccess(Response response) {
                        Toast.makeText(
                                activity,
                                "Comment posted, it might take a minute to show up",
                                Toast.LENGTH_SHORT).show();
                        onFinished.run();
                    }

                    @Override
                    public void onFailure(String summary, String response) {
                        setSubmitting(false);
                        showSubmissionFailure(summary, response, commentText);
                    }

                    @Override
                    public void onCaptchaRequired(UserActions.CaptchaChallenge challenge) {
                        UserActions.ActionCallback callback = this;
                        activity.showCaptchaDialog(
                                challenge,
                                new CaptchaResultCallback() {
                                    @Override
                                    public void onCaptchaResponse(
                                            UserActions.CaptchaChallenge captchaChallenge,
                                            String captchaResponse) {
                                        UserActions.continueCaptchaAction(
                                                activity,
                                                captchaChallenge,
                                                captchaResponse,
                                                callback);
                                    }

                                    @Override
                                    public void onCaptchaCancelled() {
                                        setSubmitting(false);
                                        Toast.makeText(
                                                activity,
                                                "Comment posting requires completing the HN captcha",
                                                Toast.LENGTH_SHORT).show();
                                    }
                                });
                    }
                });
    }

    private void setSubmitting(boolean value) {
        submitting = value;
        if (controller != null) controller.setSubmitting(value);
    }

    private void showSubmissionFailure(
            @Nullable String summary,
            @Nullable String response,
            @Nullable String commentDraft) {
        boolean isPost = type == ComposeEditorContract.TYPE_POST;
        String draftName = isPost ? "post" : "comment";
        String title;
        String message;

        if (!Utils.isNetworkAvailable(activity)) {
            title = "No internet connection";
            message = "Connect to the internet, then try again. Your " + draftName
                    + " is still here.";
        } else {
            title = isPost ? "Couldn't submit post" : "Couldn't post comment";
            StringBuilder messageBuilder = new StringBuilder();
            if (!TextUtils.isEmpty(summary)) messageBuilder.append(summary);
            if (!TextUtils.isEmpty(response)) {
                if (messageBuilder.length() > 0) messageBuilder.append("\n\n");
                messageBuilder.append(response);
            }
            if (messageBuilder.length() > 0) messageBuilder.append("\n\n");
            messageBuilder.append("Your ").append(draftName)
                    .append(" is still here, so you can edit it or try again.");
            message = messageBuilder.toString();
        }

        UserActions.showFailureDetailDialog(activity, title, message, commentDraft);
    }
}

File: app/src/main/java/com/simon/harmonichackernews/ui/editor/ComposeEditorCoordinator.kt
package com.simon.harmonichackernews.ui.editor

import androidx.lifecycle.lifecycleScope
import com.simon.harmonichackernews.MainActivity
import com.simon.harmonichackernews.harmonicAppComposition
import com.simon.harmonichackernews.app.createEditorFeatureSession
import com.simon.harmonichackernews.presentation.CaptchaResultHandler
import com.simon.harmonichackernews.ui.navigation.MainNavigationController
import com.simon.harmonichackernews.navigation.EditorDestination
import com.simon.harmonichackernews.presentation.EditorSubmission
import com.simon.harmonichackernews.presentation.EditorPresentationCopy
import com.simon.harmonichackernews.presentation.EditorWorkflowResult
import com.simon.harmonichackernews.ui.session.EditorScreenSession
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch

/** Owns submission side effects while the editor itself is a MainActivity Compose destination.  */
class ComposeEditorCoordinator(
    private val activity: MainActivity,
    destination: EditorDestination,
    private val navigation: MainNavigationController,
    private val onFinished: () -> Unit,
) {
    private val id = destination.itemId
    val type = destination.type
    val parentText: String? = destination.parentText
    val postTitle: String? = destination.postTitle
    val user: String? = destination.userName
    private var controller: EditorComposeController? = null
    private val appComposition = activity.harmonicAppComposition
    private val featureSession = appComposition.createEditorFeatureSession(
        scope = activity.lifecycleScope,
        type = type,
        itemId = id,
    )
    private val screenSession = EditorScreenSession(activity.lifecycleScope, featureSession)
    private val submittingCollectionJob: Job = activity.lifecycleScope.launch {
        screenSession.submitting.collect { controller?.updateSubmitting(it) }
    }
    private val resultCollectionJob: Job = activity.lifecycleScope.launch {
        screenSession.results.collect(::handleResult)
    }
    private var closeRequested = false
    private var resultHandlingDisposed = false
    private var awaitingWorkflowResult = false

    fun attachController(controller: EditorComposeController) {
        this.controller = controller
        controller.updateSubmitting(screenSession.submitting.value)
    }

    fun submit(submission: EditorSubmission) {
        awaitingWorkflowResult = true
        screenSession.submit(submission)
    }

    fun close() {
        if (closeRequested) return
        closeRequested = true
        controller = null
        submittingCollectionJob.cancel()
        if (!awaitingWorkflowResult) disposeResultHandling()
    }

    private fun disposeResultHandling() {
        if (resultHandlingDisposed) return
        resultHandlingDisposed = true
        resultCollectionJob.cancel()
        screenSession.dispose()
    }

    private fun handleResult(result: EditorWorkflowResult) {
        val completesPendingWorkflow = when (result) {
            is EditorWorkflowResult.Captcha -> false
            EditorWorkflowResult.Ignored -> !featureSession.isSubmitting
            EditorWorkflowResult.Success,
            is EditorWorkflowResult.Failure,
            is EditorWorkflowResult.CaptchaCancelled -> true
        }
        if (completesPendingWorkflow) awaitingWorkflowResult = false

        when (result) {
            EditorWorkflowResult.Success -> {
                navigation.showMessage(EditorPresentationCopy.successMessage(type))
                onFinished()
            }
            is EditorWorkflowResult.Failure -> navigation.showFailureDetailDialog(
                result.title,
                result.message,
                result.commentDraft,
            )
            is EditorWorkflowResult.Captcha -> navigation.showCaptchaDialog(
                result.challenge,
                object : CaptchaResultHandler {
                    override fun onCaptchaResponse(
                        challenge: com.simon.harmonichackernews.network.HackerNewsCaptchaChallenge,
                        captchaResponse: String,
                    ) {
                        screenSession.respondToCaptcha(challenge, captchaResponse)
                    }

                    override fun onCaptchaCancelled() {
                        screenSession.cancelCaptcha()
                    }
                },
            )
            is EditorWorkflowResult.CaptchaCancelled -> navigation.showMessage(result.message)
            EditorWorkflowResult.Ignored -> Unit
        }

        // Closing an idle editor releases it immediately. If a submission is already running,
        // preserve the previous behavior by handling its result (including a captcha round trip)
        // before detaching the result bridge. The feature request itself remains Activity-scoped.
        if (closeRequested && completesPendingWorkflow) disposeResultHandling()
    }
}
